src/util/test_utils.rs

```diff
@@ -0,0 +1,604 @@
+            Ok(val) => {
+                if fuzz_bytes.is_empty() {
+                    vec![]
+                } else {
+                    vec![val; (fuzz_bytes[0] % 10 + 1) as usize]
+                }
+            }
+            Err(_) => vec![H],
+        });
+
+    // SubField::MysteriousMinerGate
+    let mysterious_miner_gate_field = SubField::MysteriousMinerGate(fuzz_bytes);
+
+    let sub_fields = match add_padding {
+        AddPadding::ToFront => vec![
+            padding_field, // This will fail if the padding is not maximum
+            tx_pub_key_field,
+            nonce_field,
+            merge_mining_field,
+            additional_public_key_field,
+            mysterious_miner_gate_field,
+        ],
+        AddPadding::ToMiddle => vec![
+            tx_pub_key_field,
+            nonce_field,
+            merge_mining_field,
+            padding_field, // This will fail if the padding is not maximum
+            additional_public_key_field,
+            mysterious_miner_gate_field,
+        ],
+        AddPadding::ToRear => vec![
+            tx_pub_key_field,
+            nonce_field,
+            merge_mining_field,
+            additional_public_key_field,
+            mysterious_miner_gate_field,
+            padding_field,
+        ],
+    };
+
+    // ExtraField
+    ExtraField(sub_fields)
+}
+
+/// Fuzz for extra field's sub fields parse, called from the fuzz target
+pub fn fuzz_extra_field_parse_sub_fields(extra_field: &ExtraField, fuzz_data: &[u8]) -> bool {
+    for sub_field in &extra_field.0 {
+        let ser_sub_field = serialize(sub_field);
+        match deserialize::<SubField>(&ser_sub_field) {
+            Ok(des_sub_field) => {
+                assert_eq!(
+                    sub_field, &des_sub_field,
+                    "\nsub field: {}\nfuzz_data: {:?}",
+                    sub_field, fuzz_data
+                )
+            }
+            Err(err) => {
+                panic!(
+                    "Deserializing a serialized SubField may not fail\n({})\nsub field: {:?}\nfuzz_data: {:?}",
+                    err,
+                    sub_field,
+                    fuzz_data
+                )
+            }
+        }
+    }
+
+    true
+}
+
+/// Fuzz for extra field try parse, called from the fuzz target
+pub fn fuzz_extra_field_try_parse(
+    extra_field: &ExtraField,
+    add_padding: AddPadding,
+    fuzz_data: &[u8],
+) -> bool {
+    match RawExtraField::try_from(extra_field.clone()) {
+        Ok(raw_extra_field) => {
+            match ExtraField::try_parse(&raw_extra_field) {
+                Ok(parsed_extra_field) => {
+                    assert_eq!(
+                        extra_field, &parsed_extra_field,
+                        "\nOn 'Ok(_)\noriginal: {:?}\nparsed:   {:?}\n'fuzz_data: {:?}",
+                        extra_field, parsed_extra_field, fuzz_data
+                    )
+                }
```

Comment:
Well, variable (<255) padding in any other place than at the rear should return an error, so that it does not go into this `Ok(parsed_extra_field)` branch. This assertion will highlight if such edge cases exist (I have not seen any up to now).

Reply:
This causes a panic:
```rust
    let extra = ExtraField(vec![SubField::Padding(1), SubField::Nonce(vec![0; 8])]);
    fuzz_extra_field_try_parse(&extra, AddPadding::ToFront, &[0]);
```
The padding is consuming the whole nonce field:
```
On 'Ok(_)
original: ExtraField([Padding(1), Nonce([0, 0, 0, 0, 0, 0, 0, 0])])
parsed:   ExtraField([Padding(11)])
'fuzz_data: [0]
Left:  ExtraField([Padding(1), Nonce([0, 0, 0, 0, 0, 0, 0, 0])])
Right: ExtraField([Padding(11)])
```

Reply:
I think the behaviour is correct, please see below. **test1** should not go into the `Ok(_)` branch.

In `monero-rs` **test1** fails in the `Ok(_)` and **test2** pass in the `Ok(_)` branch:

```rust
#[test]
fn test1() {
    let extra = ExtraField(vec![SubField::Padding(1), SubField::Nonce(vec![0; 8])]);
    fuzz_extra_field_try_parse(&extra, AddPadding::ToFront, &[0]);
}
```
```
    Finished test [unoptimized + debuginfo] target(s) in 0.04s
     Running tests\fuzz_prep.rs (target\debug\deps\fuzz_prep-f8cd4a190011df06.exe)
raw_extra_field: RawExtraField([0, 0, 2, 8, 0, 0, 0, 0, 0, 0, 0, 0])

On 'Ok(_)
original: ExtraField([Padding(1), Nonce([0, 0, 0, 0, 0, 0, 0, 0])])
parsed:   ExtraField([Padding(11)])
'fuzz_data: [0]
Left:  ExtraField([Padding(1), Nonce([0, 0, 0, 0, 0, 0, 0, 0])])
Right: ExtraField([Padding(11)])
```
```rust
#[test]
fn test2() {
    let extra = ExtraField(vec![SubField::Nonce(vec![0; 8]), SubField::Padding(1)]);
    fuzz_extra_field_try_parse(&extra, AddPadding::ToFront, &[0]);
}
```
```
    Finished test [unoptimized + debuginfo] target(s) in 0.04s
     Running tests\fuzz_prep.rs (target\debug\deps\fuzz_prep-f8cd4a190011df06.exe)
raw_extra_field: RawExtraField([2, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Process finished with exit code 0
```

In `monero`, **test1** pass with assert _false_ and **test2** pass assert _true_:

```rust
TEST(hansie_custom, test1)
{
  const uint8_t extra_arr[] = {0, 0, 2, 8, 0, 0, 0, 0, 0, 0, 0, 0};
  std::vector<uint8_t> extra(&extra_arr[0], &extra_arr[0] + sizeof(extra_arr));
  std::vector<cryptonote::tx_extra_field> tx_extra_fields;
  ASSERT_FALSE(cryptonote::parse_tx_extra(extra, tx_extra_fields));
}

TEST(hansie_custom, test2)
{
  const uint8_t extra_arr[] = {2, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
  std::vector<uint8_t> extra(&extra_arr[0], &extra_arr[0] + sizeof(extra_arr));
  std::vector<cryptonote::tx_extra_field> tx_extra_fields;
  ASSERT_TRUE(cryptonote::parse_tx_extra(extra, tx_extra_fields));
}
```
```
[----------] 2 tests from hansie_custom
[ RUN      ] hansie_custom.test1
[       OK ] hansie_custom.test1 (0 ms)
[ RUN      ] hansie_custom.test2
[       OK ] hansie_custom.test2 (0 ms)
[----------] 2 tests from hansie_custom (0 ms total)
```

Reply:
Yeah your right the behavior is correct, we aren't checking if each byte is 0 in our extra field padding parsing.